lib/constructs/interface.js

```diff
@@ -416,7 +421,83 @@ class Interface {
   }
 
   generateIterator() {
-    if (this.iterable && this.iterable.isPair) {
+    if (!this.iterable) {
+      return;
+    }
+
+    if (this.iterable.isAsync) {
+      this.str += `
+        const AsyncIteratorPrototype = Object.create(utils.AsyncIteratorPrototype, {
+          [Symbol.toStringTag]: {
+            value: "${this.name} AsyncIterator",
+            configurable: true
+          }
+        });
+
+        Object.assign(AsyncIteratorPrototype, {
+          next() {
+            const internal = this[utils.iterInternalSymbol];
+            if (!internal) {
+              return Promise.reject(new TypeError("next() called on an object that is not an async iterator prototype object"));
+            }
+
+            const nextSteps = () => {
+              if (internal.isFinished) {
+                return Promise.resolve({ value: undefined, done: true });
+              }
+
+              const nextPromise = internal.target[implSymbol][utils.asyncIteratorNext](this);
+              return nextPromise.then(next => {
+                internal.ongoingPromise = null;
+                if (next === utils.asyncIteratorEOI) {
+                  internal.isFinished = true;
+                  return { value: undefined, done: true };
+                }`;
+      if (this.iterable.isPair) {
+        this.str += `
+                return utils.iteratorResult(next.map(utils.tryWrapperForImpl), kind);
+        `;
+      } else {
+        this.str += `
+                return { value: utils.tryWrapperForImpl(next), done: false };
+        `;
+      }
+      this.str += `
+              });
+            };
+
+            internal.ongoingPromise = internal.ongoingPromise ? internal.ongoingPromise.then(nextSteps) : nextSteps();
+            return internal.ongoingPromise;
+          }
+      `;
+
+      if (this.iterable.hasReturnSteps) {
+        this.str += `
+          ,
+          return(value) {
+            const internal = this[utils.iterInternalSymbol];
+            if (!internal) {
+              return Promise.reject(new TypeError("return() called on an object that is not an async iterator prototype object"));
+            }
+
+            if (internal.ongoingPromise) {
+              return Promise.reject(new TypeError("return() cannot be called while an ongoing call to next() has not settled"));
```

Comment:
Question: `ongoingPromise` is only set to `null` if the promise `next()` returns gets fulfilled, _not_ when it gets rejected. It seems to me that we might want the `[asyncIteratorReturn]` method to still be called (which probably cleans up some system resources) even if the last `next()` failed…?

Reply:
Good catch. I added a test to the streams WPT to catch this, and will send a Web IDL PR to fix it.